torch_xla_py/xla_dist.py

```diff
@@ -402,3 +409,283 @@ def get_cluster(self):
     cluster.validate()
     return cluster
 
+
+class DistributedExecutor(object):
+
+  SCRIPT_PATH_TMPL = '/tmp/{pid}/dist_training_ptxla_{worker}.sh'
+  MASTER_IDX = 0
+  MESH_SERVICE_PORT = 8477  # Use single port to disallow concurrent runs
+  DIST_ENV_VARS = [
+      'XRT_TPU_CONFIG', 'XRT_LOCAL_WORKER', 'XRT_MESH_SERVICE_ADDRESS'
+  ]
+
+  @staticmethod
+  def _parse_cmd_argument(cmd, key, flag=True):
+    args = []
+    key = '--' + key if flag else key
+    for i, part in enumerate(cmd):
+      if key == part:
+        # ex. 'docker run --name value image'
+        args.append(cmd[i + 1])
+      key_regex = re.compile('{}(=|\s)(\w*)'.format(key))
+      matches = key_regex.findall(part)
+      if len(matches) > 0:
+        # ex. If args passed in as a single string or if 'docker run
+        # --name=value image'
+        args.extend([match[1] for match in matches])
+    return args
+
+  def __init__(self, cluster):
+    self._cluster = cluster
+    logging.basicConfig(
+        format='%(asctime)-12s %(clientip)s [%(ordinal)s] %(message)s',
+        datefmt='%Y-%m-%d %H:%M:%S',
+        level=logging.DEBUG)
+    self.logger = logging.getLogger('DistributedExecutor')
+    self.container_name = None
+    self.is_docker = False
+
+  def _stream_logs(self, process, client_worker):
+    client_ip = client_worker._internal_ip
+    ordinal = self._cluster._client_workers.index(client_worker)
+
+    def _stream_output(stream, log_fn):
+      for std in iter(stream.readline, b''):
+        log_fn(
+            std.decode('utf-8').rstrip('\n'),
+            extra={
+                'clientip': client_ip,
+                'ordinal': ordinal
+            })
+
+    stdout = threading.Thread(
+        target=_stream_output, args=(
+            process.stdout,
+            self.logger.info,
+        ))
+    stdout.daemon = True
+    stdout.start()
+    stderr = threading.Thread(
+        target=_stream_output, args=(
+            process.stderr,
+            self.logger.error,
+        ))
+    stderr.daemon = True
+    stderr.start()
+    stdout.join()
+    stderr.join()
+
+  def _build_scp_cmd(self, local_path, remote_path, client_worker):
+    return [
+        'gcloud', '-q', 'compute', 'scp', '--internal-ip',
+        '--zone={}'.format(client_worker._zone), local_path,
+        'pytorchtpudistrunner@{}:~/{}'.format(client_worker._hostname,
+                                              os.path.basename(remote_path)),
+    ]
+
+  def _build_ssh_cmd(self, remote_cmd, client_worker):
+    if isinstance(remote_cmd, list):
+      remote_cmd = ' '.join(remote_cmd)
```

Comment:
I am a bit concerned around simple concatenation by space.
Something like this might be better, but I am not sure how `gcloud` command handles the escaping:
```Python
' '.join('"{}"'.format(c) for c in remote_cmd)
```